modules/siwe.ts

```diff
@@ -48,68 +50,119 @@ class Siwe {
-      nonce
-    );
+    for (let i = 0; i < addr.length; i++) {
+      if (parseInt(hash[i], 16) >= 8) {
+        ret += addr[i].toUpperCase();
+      } else {
+        ret += addr[i];
+      }
+    }
+    return ret;
+  }
 
-    console.log({ message })
+  static async signInWithEthereum(
+    discordId: string,
+    provider?: EIP1193Provider,
+    address?: string
+  ) {
+    try {
+      if (!signer) { alert('Wallet not connected!'); return }
+      let walletAddress = '';
+      if (address) {
+        walletAddress = this.encodeAddress(address);
+        // walletAddress = address;
+      } else {
+        walletAddress = await signer.getAddress();
+      }
+  
+      /* Get nonce from server */
+      const { data: nonce } = await privateBaseAxios.get(`/siwe/nonce/${walletAddress}`)
+  
+      const message = this.createSiweMessage(
+        walletAddress,
+        `Sign in with Ethereum to the app. - link to discord_id ${discordId}`,
+        nonce
+      );
+
+      // generate signature. If address is provided, use provider, otherwise use signer
+      const signature = await this.generateSignature(message, provider, address);
+  
+      // const result = this.verifySiweMessage(message, signature);
+      return await this.sendForVerification(message, signature, discordId, walletAddress);
+    } catch (e) {
+      console.error('signInWithEthereum: ', e);
+      return;
+    }
+  }
 
-    const signature = await signer.signMessage(message);
+  static async generateSignature (message: string, provider?: EIP1193Provider, address?: string) {
+    return (provider && address) ? await provider.request({ method: 'personal_sign', params: [message, address] }) : await signer.signMessage(message);
+  }
 
-    return await this.sendForVerification(message, signature, discordId);
+  static async verifySiweMessage(
+    verification_message: string,
+    signed_message: string,
+  ) {
+    try {
+      const siweMessage = new SiweMessage(verification_message);
+      const result = await siweMessage.verify({
+        signature: signed_message,
+      });
+      return siweMessage.prepareMessage();
+    } catch (e) {
+      console.error('verifySiweMessage: ', e);
+    }
   }
 
-  static createSiweMessage(address: string, statement: string, nonce: string) {
-    const siweMessage = new SiweMessage({
+  static createSiweMessage(
+    address: string,
+    statement: string,
+    nonce: string
+  ) {
+    const message = {
       domain,
       address,
       statement,
       uri: origin,
       version: '1',
       chainId: 1,
       nonce,
-
-    });
+    }
+    const siweMessage = new SiweMessage(message);
 
     return siweMessage.prepareMessage();
   }
 
-  static async sendForVerification(message: string, signature: string, discordId: string) {
+  static async sendForVerification(
+    message: string,
+    signature: string,
+    discordId: string,
+    address?: string
```

Comment:
in which case is address not known before sending for verification?

Reply:
It's not that it's not know, It's just optional.